app/src/androidTest/java/com/google/jetpackcamera/utils/UiTestUtil.kt

```diff
@@ -77,3 +90,129 @@ fun ComposeTestRule.getCurrentLensFacing(): LensFacing {
         }
     }
 }
+
+/**
+ * Rule that you to specify test methods that will have permissions granted prior to starting
+ *
+ * @param permissions the permissions to be granted
+ * @param targetTestNames the names of the tests that this rule will apply to
+ */
+class IndividualTestGrantPermissionRule(
+    private val permissions: Array<String>,
+    private val targetTestNames: Array<String>
+) :
+    TestRule {
+    private lateinit var wrappedRule: GrantPermissionRule
+
+    override fun apply(base: Statement, description: Description): Statement {
+        for (targetName in targetTestNames) {
+            if (description.methodName == targetName) {
+                wrappedRule = GrantPermissionRule.grant(*permissions)
+                return wrappedRule.apply(base, description)
+            }
+        }
+        // If no match, return the base statement without granting permissions
+        return base
+    }
+}
+
+// functions for interacting with system permission dialog
+@SdkSuppress(minSdkVersion = 30)
+fun askEveryTimeDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 30) {
+        Log.d(TAG, "Searching for Allow Once Button...")
+
+        val askPermission = findObjectById(
+            uiDevice = uiDevice,
+            resId = "com.android.permissioncontroller:id/permission_allow_one_time_button"
+        )
+
+        Log.d(TAG, "Clicking Allow Once Button")
+
+        askPermission!!.click()
+    }
+}
+
+/**
+ *  Clicks ALLOW option on an open permission dialog
+ */
+fun grantPermissionDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 23) {
+        Log.d(TAG, "Searching for Allow Button...")
+
+        val allowPermission = findObjectById(
+            uiDevice = uiDevice,
+            resId = when {
+                Build.VERSION.SDK_INT <= 29 ->
+                    "com.android.packageinstaller:id/permission_allow_button"
+                else -> "com.android.permissioncontroller:id/permission_allow_button"
```

Comment:
It's possible on some devices it could be `permission_allow_foreground_only_button`, which corresponds to "Allow while the app is running".